### ISSUE TITLE
Incorrect '$top' value generated when using 'Request.GetNextPageLink(int)'

### ISSUE THREAD

**Issue by julealgon** [OData/WebApi#1638]
When both server-driven paging and client-driven paging are involved (that is, there is a max pagesize configured on the server, and a `$top=X` is specified in the query), the generated `nextLink` value when using `GetNextPageLink(int)` is incorrect and sometimes invalid.

### Assemblies affected

Microsoft.AspNetCore.OData 7.0.1

### Reproduce steps

1. Create an OData API with server-driven paging. Set paging size to 3 (for example)
2. Configure the API to allow `$top`, max value 2 (for example)
3. Manually handle Skip/Top via `ODataQuerySettings`, and set NextLink value using `GetNextPageLink` method
3. Run query by specifying a `$top` value and observe returned `@odata.nextlink` value

Sample:
```csharp
[Page(MaxTop = 2, PageSize = 3)]
public class SampleModel
{
    public int Id { get; set; }

    public string Text { get; set; }
}

[ODataRoute("randomdata")]
public IActionResult GetData(ODataQueryOptions<SampleModel> oDataQueryOptions)
{
    var data = new[]
    {
        new SampleModel{Id = 1, Text = "First" },
        new SampleModel{Id = 2, Text = "Second" },
        new SampleModel{Id = 3, Text = "Third" },
        new SampleModel{Id = 4, Text = "Fourth" },
        new SampleModel{Id = 5, Text = "Fifth" },
        new SampleModel{Id = 6, Text = "Sixth" },
        new SampleModel{Id = 7, Text = "Seventh" },
        new SampleModel{Id = 8, Text = "Eight" },
        new SampleModel{Id = 9, Text = "Ninth" }
    };

    var pageSize = 3;
    var results = data
        .Skip(oDataQueryOptions.Skip?.Value ?? 0)
        .Take(oDataQueryOptions.Top?.Value ?? pageSize);

    Request.ODataFeature().NextLink = Request.GetNextPageLink(pageSize);

    return new OkObjectResult(results);
}
```

Request:
```http://localhost/api/randomdata?$top=2```

### Expected result

```json
{
	"@odata.context": "http://localhost/api/$metadata#randomdata",
	"value": [{
			"Id": 1,
			"Text": "First"
		}, {
			"Id": 2,
			"Text": "Second"
		}
	],
	"@odata.nextLink": "http://localhost/api/randomdata?$top=2&$skip=3"
}
```

### Actual result

(Notice the negative `$top` value)

```json
{
	"@odata.context": "http://localhost/api/$metadata#randomdata",
	"value": [{
			"Id": 1,
			"Text": "First"
		}, {
			"Id": 2,
			"Text": "Second"
		}
	],
	"@odata.nextLink": "http://localhost/api/randomdata?$top=-1&$skip=3"
}
```

### Additional detail

It appears that `GetNextPageLink` is subtracting the `PageSize` value from the `$top` counter, which doesn't make sense. I see two options here:
1. It should only modify the `$skip` counter appropriately based on the page size and leave `$top` alone
2. It should increment `$skip` and modify the `$top` value to equal `PageSize`.

EDIT:
Found the exact place in the code where this decision is made:
https://github.com/OData/WebApi/blob/d02bc61ea7b31ada1e54abbeebbecb3c5df0e3ac/src/Microsoft.AspNet.OData.Shared/GetNextPageHelper.cs#L33-L40

This doesn't make sense to me unless you are at the last page of data. I think there is a considerable gap here when dealing with `$top`/`$skip`/`PageSize` at the same time.

**Comment by julealgon** [comment]
I found a workaround to retain the `$top` value due to another problem described here:
https://github.com/OData/WebApi/issues/1639

By passing an uppercase `$TOP` value, it is still considered correctly on the API but is ignored on the nextlink generation logic.

**Comment by KanishManuja-MS** [comment]
@julealgon There is a Contract.Assert in the GetNextPageLink logic. The condition is taken care of if the EnableQuery attribute is used. Ideally, the controller method should check if the pagesize is greater than $top value for better performance if you are handling QueryOptions manually. 

I will discuss adding a safeguard to this with others and see if they think this should be fixed.

**Comment by julealgon** [comment]
> Ideally, the controller method should check if the pagesize is greater than $top value for better performance if you are handling QueryOptions manually.

@KanishManuja-MS I'm actually picking the smallest of the 2 values (client-side `$top` vs server-side `PageSize`). 

I'm not sure I follow on your reasoning there, what extra handling would I need to do? All I want is for the API to generate a sensible `@nextlink` to the consumer based on his previous input: it he requested the first X elements, the nextlink will pick up the next page of data considering the same X elements.

I really don't understand why we are decrementing the `$top` value at all here. Why would you decrement it?

**Comment by KanishManuja-MS** [comment]
@julealgon If the original request wanted top 50 records and due to server-driven paging only 10 were sent down then the nextlink should be aware that the original only requested 50; therefore only 40 more need to be sent down and that is where the decrementing from the original top comes in.

What you have here is an issue. We only throw an error in debug mode, where as this should be an exception (or handled in the code) when the top value is less than the page size.

EnableQuery attribute handles this by itself but the correct workaround for you would be only call this nextlink generation API when the number of results (or at least the top query option value) is greater than pagesize.

**Comment by julealgon** [comment]
@KanishManuja-MS 
> EnableQuery attribute handles this by itself but the correct workaround for you would be only call this nextlink generation API when the number of results (or at least the top query option value) is greater than pagesize.

Oh, I think I get it now: the `nextlink` value is only supposed to be emitted when server-side paging happens in the first place, is that it? Please confirm the scenarios below:

1. The user requests X amount of results via `$top=X`, and he **gets** the X results (when `$top` is lower than `PageSize`). A nextlink is not be emitted in the response, "because he got all that he wanted".

2. The user does not specify a `$top` value. This means "he wants all the data". When server-side paging takes place in this case we emit the `nextlink` to inform the user that "there is more data for you to fetch so that you can fulfill your original request". In the last page of data, a `@nextlink` is then not sent.

3. The user specifies a `$top` value that is greater than `PageSize`. In this case, similarly to case 2, the results will be limited by the server `PageSize` value so he needs to make another request with `$skip+=pagesize` and `$top-=pagesize`

If that's the case then I think this was more of a misunderstanding on my side on how OData is supposed to handle `@nextlink` than a problem in the library to begin with.

I'd still suggest never setting `$top` to a negative value though (as that's a completely invalid OData request to begin with). It should probably throw an exception like you mentioned in this case.

**Comment by nonbob** [comment]
summary: yes, $skiptoken is cleaner, more robust and more efficient.  My thanks to the committee and implementors for the change of direction.
tl;dr

I've read the web of links connected to this ticket and even found the relevant part of the standards doc (which I'll quote below) and I think I can see where this went off the rails (and why different interpretations are possible - until you hit the edge conditions).

Meanwhile, I'm finishing off the partial $top/$skip/offset/fetch implementation I inherited and leaving notes to myself on how to switch to $top/$skipToken/top/key >= token in the near future.

[Control Information: nextLink (odata.nextLink)](https://docs.oasis-open.org/odata/odata-json-format/v4.01/odata-json-format-v4.01.html#sec_ControlInformationnextLinkodatanextL) states:

> The nextLink control information indicates that a response is only a subset of the requested collection. It contains a URL that allows retrieving the next subset of the requested collection.  This control information can also be applied to expanded to-many navigation properties.

I took me a little while to realize that "requested collection" specifically means "the portion of the collection just requested by the client" (which may be the whole thing or just a part).  Other interpretations might be possible, but this is the only one that makes sense under all conditions.  (And there is no clarification of the phrase "requested collection" - that phrase only occurs in the section on nextLink, which I've quoted in toto above, no meta-nextLink needed)

So nextLink is an indication that the client needs to request more from the server to fulfill a specific request just made by the client; it's never a convenience to client so that the client doesn't have to construct its own link to the next page on the grid.  My first (and misguided) instinct was to think it was the latter.  Wishful thinking and laziness on my part, perhaps.  But I'm guessing I'm not the only one.
